### PR TITLE
Added vulnerability projects in cosv

### PR DESCRIPTION
**Pull request by Cheshiriks**
Снимок экрана 2023-10-17 в 18 18 22

### CODE REVIEW

File: save-frontend/src/main/resources/locales/en/proposing.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
   "Affected projects": "Affected projects",
   "Propose a vulnerability": "Propose a vulnerability",
   "Project type": "Project type",
-  "Versions": "Versions",
+  "Affected versions": "Affected versions",
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/resources/locales/en/proposing.json:17]: The key should be renamed in the Chinese version, too.

Comment (Cheshiriks) [reply save-frontend/src/main/resources/locales/en/proposing.json:17]: I deleted it so I wouldn't forget to add the correct translation later

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt [hunk 2]
```diff
@@ -21,100 +22,203 @@ import react.useState
...
+
         div {
             className = ClassName("d-sm-flex align-items-center justify-content-center mt-4")
             button {
                 type = ButtonType.button
                 className = ClassName("btn btn-outline-primary mr-3")
                 onClick = {
-                    if (vulnerabilityProject.name.isBlank() || !vulnerabilityProject.repository.isValidUrl()) {
-                        setConflictErrorMessage("Please input a valid $nameOrHash and URL".t())
+                    if (vulnerabilityProject.`package`?.name.isNullOrBlank() || !vulnerabilityProject.`package`?.repository.isValidUrl()) {
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt:217]:
```suggestion
                    val packageOrNull = vulnerabilityProject.`package`
                    if (packageOrNull?.name.isNullOrBlank() || !packageOrNull?.repository.isValidUrl()) {
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt [hunk 3]
```diff
@@ -21,100 +22,203 @@ import react.useState
...
 
-        if (vulnerabilityProject.type != VulnerabilityProjectType.COMMIT) {
             inputTextFormOptional {
-                form = InputTypes.PROJECT_VERSION
-                textValue = vulnerabilityProject.versions.joinToString(" ,")
-                validInput = conflictErrorMessage.isNullOrEmpty()
-                classes = "col-12 pl-2 pr-2 mt-3 text-left"
-                name = "Versions".t()
+                form = InputTypes.GIT_URL
+                textValue = vulnerabilityProject.`package`?.fixedCommits?.first() ?: ""
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt:191]: Please extract `vulnerabilityProject.package` into a local variable. You're using it in a number of places.

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt [hunk 4]
```diff
@@ -21,100 +22,203 @@ import react.useState
...
+                classes = "col-6 pl-2 pr-2 mt-3 text-left"
+                name = "Commits url with fix"
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt:193]: Please proofread.

At least, _URL_ (capital case) and _with **a** fix_.

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt [hunk 5]
```diff
@@ -21,100 +22,203 @@ import react.useState
...
+                name = "Introduces Url commits"
+                onChangeFun = { event ->
+                    val newPackage = vulnerabilityProject.`package`?.copy(introducedCommits = listOf(event.target.value))
+                    setVulnerabilityProject(vulnerabilityProject.copy(`package` = newPackage))
+                }
+            }
+
+            inputTextFormOptional {
+                form = InputTypes.VERSION
+                textValue = range.events.first().fixed ?: ""
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt:179]:
```suggestion
                textValue = range.events.first().fixed.orEmpty()
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt [hunk 6]
```diff
@@ -21,100 +22,203 @@ import react.useState
...
+                    setRange(newRange)
+                    setVulnerabilityProject(vulnerabilityProject.copy(ranges = listOf(newRange)))
+                }
+            }
+
+            inputTextFormOptional {
+                form = InputTypes.GIT_URL
+                textValue = vulnerabilityProject.`package`?.introducedCommits?.first() ?: ""
+                classes = "col-6 pl-2 pr-2 mt-3 text-left"
+                name = "Introduces Url commits"
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt:170]: Please be consistent about the case of abreviations.

> URL

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt [hunk 7]
```diff
@@ -21,100 +22,203 @@ import react.useState
...
-            validInput = vulnerabilityProject.name.isNotBlank()
-            classes = "col-12 pl-2 pr-2 mt-3 text-left"
-            name = "${projectTypeName.t()} ${nameOrHash.t()}"
-            onChangeFun = { event ->
-                setVulnerabilityProject(vulnerabilityProject.copy(name = event.target.value))
-                setConflictErrorMessage(null)
+            inputTextFormRequired {
+                form = InputTypes.PROJECT_NAME
+                conflictMessage = "Name must not be empty".t()
+                textValue = vulnerabilityProject.`package`?.name ?: ""
```
Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt:88]: Please extract `vulnerabilityProject.`package`?.name` into a local variable.

Comment (0x6675636b796f75676974687562) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt:88]:
```suggestion
                textValue = vulnerabilityProject.`package`?.name.orEmpty()
```

**Review by 0x6675636b796f75676974687562 — APPROVED** [review]

**Review by akuleshov7 — APPROVED** [review]

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt [hunk 8]
```diff
@@ -21,100 +22,203 @@ import react.useState
...
 import web.html.ButtonType
 
+import kotlinx.serialization.json.JsonObject
+
 /**
  * Component that allows to add new project in vulnerability
  */
 val vulnerabilityProjectWindow: FC<VulnerabilityProjectWindowProps> = FC { props ->
+
+    @Suppress("GENERIC_VARIABLE_WRONG_DECLARATION")
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt:32]: please do not suppress, fix it
val emptyRange: Range<JsonObject> = Range(

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt [hunk 9]
```diff
@@ -21,100 +22,203 @@ import react.useState
...
             }
         }
 
+        inputTextFormOptional {
+            form = InputTypes.PROJECT_VERSION
+            textValue = vulnerabilityProject.versions?.joinToString(" ,") ?: ""
+            classes = "mt-3 text-left"
+            name = "Affected versions".t()
+            onChangeFun = { event ->
+                setVulnerabilityProject(vulnerabilityProject.copy(versions = event.target.value.split(" ,")))
```
Comment (kgevorkyan) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt:207]: seems it shoulnt work

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt (hunk 2 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt]: Added

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt (hunk 3 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt (hunk 4 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt (hunk 5 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt]: Added

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt [hunk 10]
```diff
@@ -21,100 +22,203 @@ import react.useState
...
+                    val newPackage = vulnerabilityProject.`package`?.copy(introducedCommits = listOf(event.target.value))
+                    setVulnerabilityProject(vulnerabilityProject.copy(`package` = newPackage))
+                }
+            }
+
+            inputTextFormOptional {
+                form = InputTypes.VERSION
+                textValue = range.events.first().fixed ?: ""
+                classes = "col-6 pl-2 pr-2 mt-3 text-left"
+                name = "Version with fix"
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt (hunk 6 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt (hunk 7 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt]: Fixed

Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt (hunk 8 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt (hunk 9 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityProjectWindow.kt]: Changed it

**Review by kgevorkyan — APPROVED** [review]

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/vulnerability/VulnerabilityService.kt [hunk 11]
```diff
@@ -380,7 +380,7 @@ class VulnerabilityService(
                 VulnerabilityMetadataProject(
                     name = dto.name,
                     url = dto.repository,
-                    versions = dto.versions.joinToString(" ,"),
+                    versions = dto.versions.joinToString(", "),
```
Comment (nulls) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/service/vulnerability/VulnerabilityService.kt:383]: we have `com.saveourtool.save.utils.PRETTY_DELIMITER`

File: save-frontend/src/main/resources/locales/ru/proposing.json [hunk 12]
```diff
@@ -24,7 +24,6 @@
   "Commit Hash": "Хэш коммита",
   "Project Url": "URL проекта",
   "Name must not be empty": "Имя должно быть непустым",
-  "Hash must not be empty": "Хэш должен быть непустым",
   "Please input a valid name and URL": "Пожалуйста введите корректное имя и URL",
-  "Please input a valid hash and URL": "Пожалуйста введите корректный хэш и URL"
+  "Language": "Язык"
```
Comment (nulls) [inline save-frontend/src/main/resources/locales/ru/proposing.json:28]: already presented on line 10

Comment (Cheshiriks) [reply save-frontend/src/main/resources/locales/ru/proposing.json]: Fixed

**Review by nulls — APPROVED** [review]

**Pull request merged**